### DISCUSSION TITLE
Route Segment Config?

### DISCUSSION THREAD

**Discussion by offwork** [vercel/next.js#54941] [App Router] [answered]
I am quite confused about Route Segment Config.

First of all is it only used in Layout and Pages. Or should it be used even in the API's `route.ts` files? If it is used in both cases, that is, it is difficult to understand how they establish a relationship between the `page.tsx` and the `route.ts`.

Secondly, if the Route Segment Config is set on a page as follows;
```
export const dynamicParams = false // or true
export default Page({params}) {
   const {slug} = params
   //...
}
```
Is the generateStaticParams method not needed?

**Comment by devjiwonchoi** [comment]
### A1

> First of all is it only used in Layout and Pages. Or should it be used even in the API's route.ts files? If it is used in both cases, that is, it is difficult to understand how they establish a relationship between the page.tsx and the route.ts.

Yes, the `Route Segment Config` can be used on both route handlers and page components.

> The Route Segment options allows you configure the behavior of a [Page](https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts), [Layout](https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts), or [Route Handler](https://nextjs.org/docs/app/building-your-application/routing/route-handlers) by directly exporting the following variables:

It is because the route handlers (`route.ts`) do support `context.params` that are identical to the `params` in page components.

```ts
export async function GET(
  request: Request,
  { params }: { params: { slug: string } }
) {
  const slug = params.slug // 'a', 'b', or 'c'
}
```

### A2

> Secondly, if the Route Segment Config is set on a page as follows; Is the generateStaticParams method not needed?

The `dynamicParams` option is to handle the pages that are NOT using the `generateStaticParams`.
Since it is set to `true` by default, when you don't set the static params to Dynamic Routes, the routes will be generated on demand.
If you set it to `false`, the dynamic pages that do not have `genereateStaticParams` will return 404.

> false: Dynamic segments not included in generateStaticParams will return a 404.

If you have experience with the Page Router, you might be familiar with `getStaticPaths - fallback`, and this is exactly what `dynamicParams: boolean` is doing.

### Reference
- [Route Segment Config](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config)
- [dynamicParams](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#dynamicparams)

**Reply by offwork** [reply]
It's still too complicated in my opinion because NextJS's `fetch API` doesn't seem to work correctly with dynamic or static parameters.

**Comment by offwork** [comment]
Thanks for the initial reply...

Based on what you wrote, the `api` folder should actually be a reflection of the folders that represent the pages under the `app` folder. For example;
```
app/
  |_api/
    |_blog/
      |_[slug]
        route.ts
  |_blog/
    |_[slug]
      page.tsx
```

If this is the case, would it be OK to use `dynamicParams` to request from `page.tsx` to the `api` with fecth? And finally, will `dynamicParams` work properly even though the api folder structure is different from the example above?

**Reply by devjiwonchoi** [reply, marked as answer]
Yes, the `route.ts` can be placed anywhere in the route segment except on the **same segment with `page.tsx`**.

So you can't have page.tsx and route.ts on the same level.

You can also fetch from the `route.ts` to work as an API route but would recommend you not to 100% depend on them on rendering (meaning as an internal API), since the `route.ts` and corresponding `page.tsx` won't be running parallel on the build time, which will cause an error for EACCESS connection.

Short answer, yes to all of your questions.

**Comment by offwork** [comment]
thanks a lot
